### PR TITLE
Add Euler Tour Solution

### PR DESCRIPTION
**Pull request by timothygao8710**
_If any of the below doesn't apply to this Pull Request, mark the checkbox as
done._

- [x] I have tested my code
- [x] I have followed the code style guidelines mentioned
      [here](https://usaco.guide/general/adding-solution/#code-conventions)
- [x] I have properly formatted the files according to the steps mentioned
      [here](https://usaco.guide/general/adding-solution#steps)

### CODE REVIEW

File: solutions/usaco-696.mdx [hunk 1]
```diff
@@ -2,13 +2,119 @@
...
+**Time Complexity:** $\mathcal{O}(N\log N)$
+
+Recall from the [Euler tour module](/gold/tree-euler) that each subtree is represented as a contiguous interval in its Euler tour. With this knowledge, we can compress the tree into an array. Now, the problem becomes for each node $i$, look for how many elements are less than $\texttt{tour}[i]$ from $i$ to $\texttt{tout}[i]$, where $\texttt{tout}[i]$ is the end of the subtree interval for node $i$. By processing the nodes in decreasing order, this is a problem we can solve in $\mathcal{O}(N \log N)$ time with a binary-indexed tree.
+<LanguageSection>
+<JavaSection>
+
+```java
+import java.io.*;
+import java.util.ArrayList;
+import java.util.Arrays;
```
Comment (nathangong) [inline solutions/usaco-696.mdx:21]:
```suggestion
import java.util.*;
```

File: solutions/usaco-696.mdx [hunk 2]
```diff
@@ -2,13 +2,119 @@
...
+
+public class promote {
+
```
Comment (nathangong) [inline solutions/usaco-696.mdx:24]:
```suggestion
public class PromotionCounting {
```

File: solutions/usaco-696.mdx [hunk 3]
```diff
@@ -2,13 +2,119 @@
...
 ---
 
-## Merging Indexed Sets
+[Official Editorial (Java)](http://www.usaco.org/current/data/sol_promote_platinum_jan17.html)
+
+## Solution 1: Euler Tour
+
+**Time Complexity:** $\mathcal{O}(N\log N)$
+
+Recall from the [Euler tour module](/gold/tree-euler) that each subtree is represented as a contiguous interval in its Euler tour. With this knowledge, we can compress the tree into an array. Now, the problem becomes for each node $i$, look for how many elements are less than $\texttt{tour}[i]$ from $i$ to $\texttt{tout}[i]$, where $\texttt{tout}[i]$ is the end of the subtree interval for node $i$. By processing the nodes in decreasing order, this is a problem we can solve in $\mathcal{O}(N \log N)$ time with a binary-indexed tree.
```
Comment (nathangong) [inline solutions/usaco-696.mdx:14]: "the problem becomes for each node $i$..." - grammatical error

**Review by nathangong — CHANGES_REQUESTED** [review]
Mostly optional changes

File: solutions/usaco-696.mdx [hunk 4]
```diff
@@ -2,13 +2,119 @@
...
 source: USACO Platinum 2017 January
 title: Promotion Counting
-author: Benjamin Qi
+author: Benjamin Qi, Timothy Gao
 ---
 
-## Merging Indexed Sets
+[Official Editorial (Java)](http://www.usaco.org/current/data/sol_promote_platinum_jan17.html)
+
+## Solution 1: Euler Tour
```
Comment (nathangong) [inline solutions/usaco-696.mdx:10]: maybe also include PURS in the solution title

File: solutions/usaco-696.mdx [hunk 5]
```diff
@@ -2,13 +2,119 @@
...
+		}
+
+		public void update(int idx, int v) {
+			idx++;
+			while (idx < bit.length) {
+				bit[idx] += v;
+				idx += idx & -idx;
+			}
+		}
+	}
```
Comment (Aadit-Ambadkar) [inline solutions/usaco-696.mdx:105]: Wrap this in a codesnip

**Review by Aadit-Ambadkar — CHANGES_REQUESTED** [review]

File: solutions/usaco-696.mdx (hunk 5 above)
Comment (nathangong) [reply solutions/usaco-696.mdx:105]: Not sure if we normally do this since it's a data structure

File: solutions/usaco-696.mdx (hunk 3 above)
Comment (timothygao8710) [reply solutions/usaco-696.mdx]: thanks, fixed

File: solutions/usaco-696.mdx [hunk 6]
```diff
@@ -2,13 +2,117 @@
...
 ---
 
-## Merging Indexed Sets
+[Official Editorial (Java)](http://www.usaco.org/current/data/sol_promote_platinum_jan17.html)
+
+## Solution 1: Euler Tour
+
+**Time Complexity:** $\mathcal{O}(N\log N)$
+
+Recall from the [Euler tour module](/gold/tree-euler) that each subtree is represented as a contiguous interval in its Euler tour. With this knowledge, we can compress the tree into an array. Now, the problem is reduced the following: For each node $i$, we need to know how many elements are less than $\texttt{tour}[i]$ from $i$ to $\texttt{tout}[i]$, where $\texttt{tout}[i]$ is the end of the subtree interval for node $i$. By processing the nodes in decreasing order, this is a problem we can solve in $\mathcal{O}(N \log N)$ time with a binary-indexed tree.
```
Comment (nathangong) [inline solutions/usaco-696.mdx:14]:
```suggestion
Recall from the [Euler tour module](/gold/tree-euler) that each subtree is represented as a contiguous interval in its Euler tour. With this knowledge, we can compress the tree into an array. Now, the problem is reduced the following: for each node $i$, we need to determine the number of elements that are less than $\texttt{tour}[i]$ from $i$ to $\texttt{tout}[i]$, where $\texttt{tout}[i]$ is the end of the subtree interval for node $i$. By processing the nodes in decreasing order, this is a problem we can solve in $\mathcal{O}(N \log N)$ time with a binary-indexed tree.
```

**Comment by dutinmeow** [comment]
that's alot of commits

File: solutions/usaco-696.mdx (hunk 5 above)
Comment (dutinmeow) [reply solutions/usaco-696.mdx:103]: It's not that long in the first place

**Review by dutinmeow — APPROVED** [review]
see changes above

**Review by dutinmeow — COMMENTED** [review]
i already approved?

**Comment by timothygao8710** [comment]
lol ye mb

**Comment by timothygao8710** [comment]
@Aadit-Ambadkar @nathangong approval?

File: solutions/usaco-696.mdx [hunk 7]
```diff
@@ -2,13 +2,117 @@
...
+import java.io.*;
+import java.util.*;
+
+public class PromotionCounting {
+	public static void main(String[] args) throws IOException {
+		BufferedReader br = new BufferedReader(new FileReader("promote.in"));
+		PrintWriter pw = new PrintWriter(new BufferedWriter(new FileWriter("promote.out")));
+	 
+		int N = Integer.parseInt(br.readLine());
+		int[][] ratings = new int[N][2];
```
Comment (nathangong) [inline solutions/usaco-696.mdx:28]: might be better to make this a class for it to be more readable, but its probably fine since this is in plat already

**Review by nathangong — APPROVED** [review]

File: solutions/usaco-696.mdx (hunk 7 above)
Comment (timothygao8710) [reply solutions/usaco-696.mdx:28]: 👍

**Comment by timothygao8710** [comment]
@Aadit-Ambadkar yo?

**Comment by Aadit-Ambadkar** [comment]
oops yeah gimme a minute or 2

File: solutions/usaco-696.mdx [hunk 8]
```diff
@@ -2,13 +2,117 @@
...
+		for (int i = 0; i < N; i++) {
+			ratings[i] = new int[]{i, Integer.parseInt(br.readLine())};
```
Comment (Aadit-Ambadkar) [inline solutions/usaco-696.mdx:30]:
```suggestion
			ratings[i][0] = i;
			ratings[i][1] = Integer.parseInt(br.readLine());
```

**Review by Aadit-Ambadkar — APPROVED** [review]
tbh the dfs probably is inefficient but ig its ok. just this one change (accepting cause its pretty small, pls make the change)

File: solutions/usaco-696.mdx (hunk 8 above)
Comment (bqi343) [reply solutions/usaco-696.mdx]: what's wrong with this?

**Pull request merged**